### PR TITLE
[delivers #177410433] nodejs-winrm package should support wsman enumerate

### PR DESCRIPTION
**Pull request by brantwedel**
Added enumerate support to nodejs-winrm package to more efficiently poll WMI information, and avoid memory leaks.

Branch also includes existing agent/keep-alive changes.

### CODE REVIEW

File: src/enumerate.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            'wsen:EnumerationContext': _params.enumerationId
+        }]
+    };
+
+    return js2xmlparser.parse('s:Envelope', res);
+}
+
+function getObjects(items) {
+    let itemObjects = [];
+    if (items) {
```
Comment (gavinaiken) [inline src/enumerate.js]: why not make this an early return ie
```
if (!items) {
    return;
}
```

File: src/enumerate.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        for (let item of items[0][Object.keys(items[0])[0]]) {
+            let itemObject = {};
+            // TODO return map of multiple item keys?
+            Object.keys(item).forEach((k) => {
```
Comment (gavinaiken) [inline src/enumerate.js]: why mix `for` and `forEach` here - nicer to stick to one or the other for both loops?

File: src/enumerate.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        }]
+    };
+
+    return js2xmlparser.parse('s:Envelope', res);
+}
+
+function getObjects(items) {
+    let itemObjects = [];
+    if (items) {
+        for (let item of items[0][Object.keys(items[0])[0]]) {
```
Comment (gavinaiken) [inline src/enumerate.js]: `items[0][Object.keys(items[0])[0]]` could be a bit more readable...

File: src/enumerate.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    var auth = _params.auth;
+    if (_params.authOnce) {
+        auth = typeof _params.authOnce === 'string' ? _params.authOnce : _params.auth;
+        _params.auth = undefined;
+        _params.authOnce = undefined;
+    }
+    var result = await winrm_http_req.sendHttp(req, _params.host, _params.port, _params.path, auth, _params.agent);
+
+    if (result['s:Envelope']['s:Body'][0]['s:Fault']) {
+        var faultCode = result['s:Envelope']['s:Body'][0]['s:Fault'][0]['s:Code'][0]['s:Subcode'][0]['s:Value'][0];
```
Comment (gavinaiken) [inline src/enumerate.js]: are these structures always guaranteed to be correct? looks like these could be prone to a 'can't read property ? of undefined` error...

Comment (brantwedel) [reply src/enumerate.js]: I think `s:Subcode` should always exist for this, but yep, the spec says `s:Reason` is optional. I'll check them both and return something regardless.

**Review by gavinaiken — APPROVED** [review]
:shipit:

**Pull request merged**